src/pages/index.js

```diff
@@ -80,6 +56,165 @@ const cardLinkInput = addCardFormElement.querySelector(
 );
 const cardSelector = "#card-template";
 
+// Likes Api
+
+function handleCardLikeClick(cardId, isLiked) {
+  if (isLiked) {
+    api
+      .unlikeCard(cardId)
+      .then((cardData) => {
+        this._likes = cardData.isLiked;
+        this._renderLikes();
+      })
+      .catch((err) => {
+        console.error(err);
+      });
+  } else {
+    api
+      .likeCard(cardId)
+      .then((cardData) => {
+        this._likes = cardData.isLiked;
+        this._renderLikes();
+      })
+      .catch((err) => {
+        console.error(err);
+      });
+  }
+}
+
+// Edit Avatar
+
+const editAvatarPopup = new PopupWithForm(
+  "#avatar-edit-modal",
+  handleEditAvatar,
+  "Save"
+);
+
+function handleEditAvatar(data) {
+  editAvatarPopup.setLoading(true);
+  api
+    .updateProfileAvatar(data.link)
+    .then((userData) => {
+      userInfo.setUserAvatar(userData.avatar);
+      editAvatarPopup.close();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      editAvatarPopup.setLoading(false, "Saving");
+    });
+}
+
+editAvatarPopup.setEventListeners();
+
+profileAvatarButton.addEventListener("click", () => editAvatarPopup.open());
+
+// User Info + Update Profile
+
+const userInfo = new UserInfo(profileTitle, profileDescription, avatarElement);
+
+function fillProfileForm() {
+  const userInfoData = userInfo.getUserInfo();
+  profileTitleInput.value = userInfoData.name;
+  profileDescriptionInput.value = userInfoData.description;
+}
+
+api.getUserInfo().then((userData) => {
+  userInfo.setUserInfo({
+    name: userData.name,
+    description: userData.about,
+  });
+});
+
+function handleProfileEditSubmit(data) {
+  editProfilePopup.setLoading(true);
+  api
+    .updateProfile(data)
+    .then((userData) => {
+      userInfo.setUserInfo({
+        name: userData.name,
+        description: userData.about,
+      });
+      fillProfileForm();
+      editProfilePopup.close();
+    })
+    .catch((error) => {
+      console.error(error);
+    })
+    .finally(() => {
+      editProfilePopup.setLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```